gitlab_test.go

```diff
@@ -0,0 +1,115 @@
+package gitlab
+
+import (
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"reflect"
+	"strings"
+	"testing"
+)
+
+var (
+// mux is the HTTP request multiplexer used with the test server.
+	mux *http.ServeMux
+
+// client is the Gitlab client being tested.
+	client *Client
+
+// server is a test HTTP server used to provide mock API responses.
+	server *httptest.Server
+)
+
+// setup sets up a test HTTP server along with a gitlab.Client that is
+// configured to talk to that test server.  Tests should register handlers on
+// mux which provide mock responses for the API method being tested.
+func setup() {
+	// test server
+	mux = http.NewServeMux()
+	server = httptest.NewServer(mux)
+
+	// gitlab client configured to use test server
+	client = NewClient(nil, "")
+	client.SetBaseURL(server.URL)
+}
+
+// teardown closes the test HTTP server.
+func teardown() {
+	server.Close()
+}
+
+func testMethod(t *testing.T, r *http.Request, want string) {
+	if got := r.Method; got != want {
+		t.Errorf("Request method: %v, want %v", got, want)
+	}
+}
+
+type values map[string]string
+
+func testFormValues(t *testing.T, r *http.Request, values values) {
+	want := url.Values{}
+	for k, v := range values {
+		want.Add(k, v)
+	}
+
+	r.ParseForm()
+	if got := r.Form; !reflect.DeepEqual(got, want) {
+		t.Errorf("Request parameters: %v, want %v", got, want)
+	}
+}
+
+func testHeader(t *testing.T, r *http.Request, header string, want string) {
+	if got := r.Header.Get(header); got != want {
+		t.Errorf("Header.Get(%q) returned %s, want %s", header, got, want)
+	}
+}
+
+func testBody(t *testing.T, r *http.Request, want string) {
+	b, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		t.Errorf("Error reading request body: %v", err)
+	}
+	if got := string(b); got != want {
+		t.Errorf("request Body is %s, want %s", got, want)
+	}
+}
+
+func responseBody(w http.ResponseWriter, filename string) {
+	body, _ := ioutil.ReadFile(filename)
+	w.Write([]byte(body))
+}
+
+func TestNewClient(t *testing.T) {
+	c := NewClient(nil, "")
+
+	if got, want := c.BaseURL().String(), defaultBaseURL; got != want {
```

Comment:
I think you've gotten a little overboard with the `got`, `want` part :wink: This is just as clean if not cleaner (IMO):

 ```go
if c.BaseURL().String() != defaultBaseURL {
  t.Errorf("NewClient BaseURL is %s, want %s", c.BaseURL().String(), defaultBaseURL)
}
```

Also notice the `%s` instead of the `%v`. We get a string here so let's format it as one. `%v` has a different use case and can be a saver option when formatting and error (`err` is not always a standard error everywhere). But if we are getting a string please use the proper formatting option.